lib/spack/spack/url_buildcache.py

```diff
@@ -0,0 +1,1129 @@
+            # spec url, or else raise an exception
+            self.read_manifest(verify_signature=not allow_unsigned)
+
+        local_specfile_path = self.fetch_blob(self.get_blob_record(BuildcacheComponent.SPEC))
+
+        # Check spec file for validity and read it, or else cleanup and exit early
+        try:
+            spec_dict, _ = get_valid_spec_file(local_specfile_path, self.get_layout_version())
+        except InvalidMetadataFile as e:
+            self.destroy()
+            raise BuildcacheEntryError("Buildcache entry does not have valid metadata file") from e
+
+        return spec_dict
+
+    def fetch_archive(self, allow_unsigned: bool = False) -> str:
+        """Retrieve the archive file and return the local archive file path"""
+        if not self.manifest:
+            # Raises if problems encountered, including not being able to verify signagure
+            self.read_manifest(verify_signature=not allow_unsigned)
+
+        return self.fetch_blob(self.get_blob_record(BuildcacheComponent.TARBALL))
+
+    def get_archive_stage(self) -> Optional[spack.stage.Stage]:
+        return self.stages[self.get_blob_record(BuildcacheComponent.TARBALL)]
+
+    def remove(self):
+        if self.manifest:
+            try:
+                web_util.remove_url(self.remote_manifest_url)
+            except Exception as e:
+                tty.debug(f"Failed to remove previous manfifest: {e}")
+
+            try:
+                web_util.remove_url(
+                    self.get_blob_url(self.get_blob_record(BuildcacheComponent.TARBALL))
+                )
+            except Exception as e:
+                tty.debug(f"Failed to remove previous archive: {e}")
+
+            try:
+                web_util.remove_url(
+                    self.get_blob_url(self.get_blob_record(BuildcacheComponent.SPEC))
+                )
+            except Exception as e:
+                tty.debug(f"Failed to remove previous metadata: {e}")
+
+            self.manifest = None
+
+    @classmethod
+    def push_blob(cls, mirror_url: str, blob_path: str, record: BlobRecord) -> None:
+        blob_destination_url = url_util.join(
+            mirror_url,
+            *cls.get_relative_path_components(BuildcacheComponent.BLOBS),
+            record.checksum_alg,
+            record.checksum[:2],
+            record.checksum,
+        )
+
+        web_util.push_to_url(blob_path, blob_destination_url, keep_original=False)
+
+    @classmethod
+    def push_manifest(
+        cls,
+        mirror_url: str,
+        manifest_name: str,
+        manifest: BuildcacheManifest,
+        tmpdir: str,
+        signing_key: Optional[str] = None,
+    ) -> None:
+        # write the manifest to a temporary location
+        manifest_file_name = f"{manifest_name}.manifest.json"
+        manifest_path = os.path.join(tmpdir, manifest_file_name)
+        with open(manifest_path, "w", encoding="utf-8") as f:
+            json.dump(manifest.to_json(), f)
```

Comment:
```suggestion
            # Note: when using gpg clear sign, we need to avoid long lines (19995 chars).
            # If lines are longer, they are truncated without error.
            # So, here we still add newlines, but no indent, so save on file size and
            # line length.
            json.dump(manifest.to_json(), f, indent=0, separators=(",", ":"))
```

Reply:
These manifests are tiny compared to the spec files we used to clearsign, so I don't think we'll ever get close to 20K chars.  But I agree there's no harm in using the `indent` and `separators` params as you used to do with specs, I'll fix that.